cajita/src/Cajita_FastFourierTransform.hpp

```diff
@@ -23,291 +23,509 @@
+     \param ScaleType Method of scaling data.
     */
-    template <class Array_t>
-    void reverse( const Array_t& x )
+    template <class Array_t, class ScaleType>
+    void reverseImpl( const Array_t& x, const ScaleType )
     {
-        compute( x, -1 );
+        compute( x, -1, Impl::HeffteScalingTraits<ScaleType>().scaling_type );
     }
 
-  public:
-    template <class Array_t>
-    void compute( const Array_t& x, const int flag )
+    /*!
+     \brief Copy data from Kokkos::complex to CUDA complex.
+     \param x_view_val Kokkos::complex value.
+     \param work_view_val CUDA complex value.
+    */
+    template <class ComplexType>
+    KOKKOS_INLINE_FUNCTION ComplexType copyFromKokkosComplex(
+        Kokkos::complex<value_type> x_view_val, ComplexType work_view_val,
+        typename std::enable_if<( is_cuda_complex<ComplexType>::value ),
+                                int>::type* = 0 )
     {
-        static_assert( is_array<Array_t>::value, "Must use an array" );
-        static_assert(
-            std::is_same<typename Array_t::entity_type, entity_type>::value,
-            "Array entity type mush match transform entity type" );
-        static_assert(
-            std::is_same<typename Array_t::mesh_type, mesh_type>::value,
-            "Array mesh type mush match transform mesh type" );
-        static_assert(
-            std::is_same<typename Array_t::device_type, DeviceType>::value,
-            "Array device type and transform device type are different." );
-        static_assert(
-            std::is_same<typename Array_t::value_type,
-                         Kokkos::complex<value_type>>::value ||
-                std::is_same<typename Array_t::value_type, value_type>::value,
-            "Array value type and complex transform value type are "
-            "different." );
+        work_view_val.x = x_view_val.real();
+        work_view_val.y = x_view_val.imag();
+        return work_view_val;
+    }
+    /*!
+     \brief Copy data from CUDA complex to Kokkos complex.
+     \param work_view_val CUDA complex value.
+     \param x_view_val Kokkos complex value.
+    */
+    template <class ComplexType>
+    KOKKOS_INLINE_FUNCTION Kokkos::complex<value_type> copyToKokkosComplex(
+        ComplexType work_view_val, Kokkos::complex<value_type> x_view_val,
+        typename std::enable_if<( is_cuda_complex<ComplexType>::value ),
+                                int>::type* = 0 )
+    {
+        x_view_val.real() = work_view_val.x;
+        x_view_val.imag() = work_view_val.y;
+        return x_view_val;
+    }
 
-        if ( 1 != x.layout()->dofsPerEntity() )
-            throw std::logic_error(
-                "Only 1 complex value per entity allowed in FFT" );
+    /*!
+     \brief Copy data from Kokkos::complex to std::complex.
+     \param x_view_val Kokkos::complex value.
+     \param work_view_val std::complex value.
+    */
+    template <class ComplexType>
+    ComplexType copyFromKokkosComplex(
```

Comment:
These need to be `KOKKOS_INLINE_FUNCTION`. Even if `std::complex` won't run on device we want it to be `inline`.

Reply:
I used `inline` since it can't be called from the device